### PR TITLE
PHP 8.1 compatibility: replace FILTER_SANITIZE_STRING

### PR DESCRIPTION
**Pull request by sjinks**
PHP 8.1 [deprecated](https://www.php.net/manual/en/filter.filters.sanitize.php) `FILTER_SANITIZE_STRING`. This PR replaces it with other alternatives.

Ref: Automattic/vip-go-mu-plugins#2667

### CODE REVIEW

File: class-two-factor-core.php [hunk 1]
```diff
@@ -537,8 +537,8 @@ public static function show_two_factor_login( $user ) {
 	 */
 	public static function backup_2fa() {
 		$wp_auth_id = filter_input( INPUT_GET, 'wp-auth-id', FILTER_SANITIZE_NUMBER_INT );
-		$nonce      = filter_input( INPUT_GET, 'wp-auth-nonce', FILTER_SANITIZE_STRING );
-		$provider   = filter_input( INPUT_GET, 'provider', FILTER_SANITIZE_STRING );
+		$nonce      = filter_input( INPUT_GET, 'wp-auth-nonce', FILTER_CALLBACK, [ 'options' => 'sanitize_key' ] );
+		$provider   = filter_input( INPUT_GET, 'provider', FILTER_CALLBACK, [ 'options' => 'sanitize_text_field' ] );
```
Comment (sjinks) [inline class-two-factor-core.php:541]: `$provider` can contain uppercase letters; therefore, `sanitize_key()` is not suitable here.

File: providers/class-two-factor-backup-codes.php [hunk 2]
```diff
@@ -304,8 +304,8 @@ public function authentication_page( $user ) {
 	 * @return boolean
 	 */
 	public function validate_authentication( $user ) {
-		$backup_code = isset( $_POST['two-factor-backup-code'] ) ? sanitize_text_field( wp_unslash( $_POST['two-factor-backup-code'] ) ) : false;
-		return $this->validate_code( $user, filter_var( $backup_code, FILTER_SANITIZE_STRING ) );
+		$backup_code = isset( $_POST['two-factor-backup-code'] ) ? sanitize_text_field( wp_unslash( $_POST['two-factor-backup-code'] ) ) : '';
+		return $this->validate_code( $user, $backup_code );
```
Comment (sjinks) [inline providers/class-two-factor-backup-codes.php:308]: `$backup_code` is already sanitized with `sanitize_text_field()`; `filter_var()` is not really needed. To ensure that `$backup_code` is a string, I replaced `false` with an empty string.

**Comment by jeffpaul** [comment]
Per [yesterday's bug scrub](https://wordpress.slack.com/archives/C04L4KCUB/p1648040434425739), @georgestephanis feels the sanitize string is reasonable and could be good to merge, but is not familiar with `filter_callback` and not positive if core code style still recommends using `array( )` over `[ ]` shorthand.

File: class-two-factor-compat.php [hunk 3]
```diff
@@ -35,7 +35,7 @@ public function init() {
 	 * @return boolean
 	 */
 	public function jetpack_rememberme( $rememberme ) {
-		$action = filter_input( INPUT_GET, 'action', FILTER_SANITIZE_STRING );
+		$action = filter_input( INPUT_GET, 'action', FILTER_CALLBACK, [ 'options' => 'sanitize_key' ] );
```
Comment (GaryJones) [inline class-two-factor-compat.php:38]:
```suggestion
		$action = filter_input( INPUT_GET, 'action', FILTER_CALLBACK, array( 'options' => 'sanitize_key' ) );
```

File: class-two-factor-core.php [hunk 4]
```diff
@@ -244,7 +244,7 @@ public static function get_user_update_action_url( $user_id, $action ) {
 	 * @return boolean
 	 */
 	public static function is_valid_user_action( $user_id, $action ) {
-		$request_nonce = filter_input( INPUT_GET, self::USER_SETTINGS_ACTION_NONCE_QUERY_ARG, FILTER_SANITIZE_STRING );
+		$request_nonce = filter_input( INPUT_GET, self::USER_SETTINGS_ACTION_NONCE_QUERY_ARG, FILTER_CALLBACK, [ 'options' => 'sanitize_key' ] );
```
Comment (GaryJones) [inline class-two-factor-core.php:247]:
```suggestion
		$request_nonce = filter_input( INPUT_GET, self::USER_SETTINGS_ACTION_NONCE_QUERY_ARG, FILTER_CALLBACK, array( 'options' => 'sanitize_key' ) );
```

File: class-two-factor-core.php [hunk 5]
```diff
@@ -277,7 +277,7 @@ public static function current_user_being_edited() {
 	 * @return void
 	 */
 	public static function trigger_user_settings_action() {
-		$action  = filter_input( INPUT_GET, self::USER_SETTINGS_ACTION_QUERY_VAR, FILTER_SANITIZE_STRING );
+		$action  = filter_input( INPUT_GET, self::USER_SETTINGS_ACTION_QUERY_VAR, FILTER_CALLBACK, [ 'options' => 'sanitize_key' ] );
```
Comment (GaryJones) [inline class-two-factor-core.php:280]:
```suggestion
		$action  = filter_input( INPUT_GET, self::USER_SETTINGS_ACTION_QUERY_VAR, FILTER_CALLBACK, array( 'options' => 'sanitize_key' ) );
```

File: class-two-factor-core.php (hunk 1 above)
Comment (GaryJones) [inline class-two-factor-core.php:541]:
```suggestion
		$nonce      = filter_input( INPUT_GET, 'wp-auth-nonce', FILTER_CALLBACK, array( 'options' => 'sanitize_key' ) );
		$provider   = filter_input( INPUT_GET, 'provider', FILTER_CALLBACK, array( 'options' => 'sanitize_text_field' ) );
```

File: class-two-factor-core.php [hunk 6]
```diff
@@ -794,7 +794,7 @@ public static function verify_login_nonce( $user_id, $nonce ) {
 	 */
 	public static function login_form_validate_2fa() {
 		$wp_auth_id = filter_input( INPUT_POST, 'wp-auth-id', FILTER_SANITIZE_NUMBER_INT );
-		$nonce      = filter_input( INPUT_POST, 'wp-auth-nonce', FILTER_SANITIZE_STRING );
+		$nonce      = filter_input( INPUT_POST, 'wp-auth-nonce', FILTER_CALLBACK, [ 'options' => 'sanitize_key' ] );
```
Comment (GaryJones) [inline class-two-factor-core.php:797]:
```suggestion
		$nonce      = filter_input( INPUT_POST, 'wp-auth-nonce', FILTER_CALLBACK, array( 'options' => 'sanitize_key' ) );
```

File: class-two-factor-core.php [hunk 7]
```diff
@@ -810,7 +810,7 @@ public static function login_form_validate_2fa() {
 			exit;
 		}
 
-		$provider = filter_input( INPUT_POST, 'provider', FILTER_SANITIZE_STRING );
+		$provider = filter_input( INPUT_POST, 'provider', FILTER_CALLBACK, [ 'options' => 'sanitize_text_field' ] );
```
Comment (GaryJones) [inline class-two-factor-core.php:813]:
```suggestion
		$provider = filter_input( INPUT_POST, 'provider', FILTER_CALLBACK, array( 'options' => 'sanitize_text_field' ) );
```

**Review by GaryJones — COMMENTED** [review]
> not familiar with filter_callback

`FILTER_CALLBACK` is [documented here](https://www.php.net/manual/en/filter.filters.misc.php), but essentially is a way to ignore the built-in `FILTER_*` constants, and set a callback function instead.

> not positive if core code style still recommends using array( ) over [ ] shorthand

WPCS still requires longhand array syntax, so that's the code suggestions I've made here.

**Review by GaryJones — APPROVED** [review]

**Comment by kasparsd** [comment]
This looks good! I'll merge this in and some linter checks to find any other PHP8 compatibility issues.

**Comment by iandunn** [comment]
> WPCS still requires longhand array syntax, so that's the code suggestions I've made here.

There was [a proposal](https://make.wordpress.org/core/2019/03/26/coding-standards-updates-for-php-5-6/) to change to the short syntax, but [the decision](https://make.wordpress.org/core/2019/07/12/php-coding-standards-changes/) was to keep the long one.

**Pull request merged**